doc/SAI-Proposal-SAI-PTF.md

```diff
@@ -0,0 +1,132 @@
+# SAI PTF
+
+| Title       | SAI PTF         |
+| ----------- | --------------- |
+| Authors     | Intel           |
+| Status      | In review       |
+| Type        | Standards track |
+| Created     | 07/10/2021      |
+| SAI-Version | 1.9             |
+
+## Overview
+
+SAI PTF (Packet Testing Framework) is an autogenerated Python based dataplane testing framework. It is using thrift features to call C-based SAI functions in Python scripts. The main goal of the framework is to reduce the time spent on writing wrapper functions for SAI which could be generated based on SAI headers.
+
+Apart from the framework there's a set of tests separated in per-feature files that are using SAI PTF to verify correctness of SAI layer implemented on top of any switch.
+
+## Autogeneration
+
+Autogeneration part, written in Perl, is using SAI headers to create necessary code to test SAI functionality.
+
+The files that are created during autogeneration are:
+
+- **sai.thrift** - containing thrift interface definitions
+- **sai_rpc_server.cpp** - a set of functions called via thrift from sai_adapter, which later calls SAI functions
+- **sai_adapter.py** - python wrappers for calling SAI functions
+
+These files should be generated once, per every SAI version that needs to be tested.
+
+| Input files                                                  | Output files                                                 |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| **Templates** <br /> - *sai.thrift.tt* <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai.thrit** <br /> - thrift RPC "header" <br /> - generated from scratch <br /> - needed by thrift to generate *sai_rcp_server.skeleton.cpp* <br /> - skeleton is used to generate *sai_rpc_server.cpp.tt* template |
+| **Templates** <br /> - *sai_rpc_server_functions.tt* <br /> - *sai_rpc_server_helper_functions.tt* <br /><br />**Auto-generated templates** <br /> - *sai_rpc_server.cpp.tt* (by Thrift from *sai.thift*) <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_rpc_server.cpp** <br /> - thrift server <br /> - need *sai.thrift* and *Thrift* itself for generation |
+| **Templates** <br /> - *sai_adapter.py.tt* <br /> - *sai_adapter_utils.tt* <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_adapter.py** <br /> - thrift client and main SAI PTF test library <br /> - generated from scratch |
+
+
+All of this can be generated with gensairpc.pl script.
+
+Autogeneration:
+
+```
+cd <root>/sai/
+./gensairpc.pl --clean
+```
+
+When upgrading SAI:
+
+```
+cd SAI/meta
+make
+cp saimetadata.c <root>/sai
+cp saimetadata.h <root>/sai
+cd <root>/sai
+./gensairpc.pl --clean --no-meta-build
+make clean - C SAI/meta
+```
```

Comment:
Can we have a simple flow diagram for the autogeneration and how SAI PTF scripts are using the thrift APIs to call C-APIs to program the SAI adapter?